### PR TITLE
Try: Create a site-card that can be iframed

### LINKED ISSUE #45574: Nav Unification: Notices and Nudges
All nudges and notices are concentrated in the [`CurrentSite` component](https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/current-site/index.jsx). This component controls everything that shows above the standard nav links in Calypso sidebar.

Screenshot 2020-09-10 at 12 36 34

## ✅ Implemented — [The Site Switcher link](https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/current-site/index.jsx#L69-L78)
- ONLY the link itself; the switching logic and UI lives elsewhere
- on click, it calls an internal Calypso function `setLayoutFocus( 'sites' )` which controls the "sidebar routing" and makes it show and focus the site list

## ✅ Implemented simplified version — [Site card](https://github.com/Automattic/wp-calypso/tree/master/client/blocks/site)
- shows the name and icon of the site)
- clicking it will open the "View Site" route — /view/:slug

## ✅ implemented [Domain Credit](https://github.com/Automattic/wp-calypso/issues/52368)
- says "Free domain available" via the `UpsellNudge` component
- eligibility determined from the redux store `state/sites/plans`

## ✅ implemented [Domain Upsell](https://github.com/Automattic/wp-calypso/issues/52368)
- different formulations and recommended tld and price based on different redux state checks
- `UpsellNudge` component

## ✅ implemented [JITM (Just In Time Messages) for wp-admin](https://github.com/Automattic/wp-calypso/issues/52368)
- This is a newer and different way to show notices in Calypso and wp-admin (https://github.com/Automattic/wp-calypso/tree/master/client/blocks/jitm)
- Uses redux state connected to `/jetpack/v4/jitm` API that returns messages and determines when to show them
- Renders using `UpsellNudge`
- I haven't checked the wp-admin side of this (likely in Jetpack). If it works, it will be the first and only notice that has an existing wp-admin representation

### [Domain Warnings](https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/current-site/domain-warnings.jsx)
- 11 different warnings
- List of possible types: https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/current-site/domain-warnings.jsx#L21-L33
- Visual representation and visibility conditions: https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/domains/components/domain-warnings/index.jsx
- Uses redux data store and the wpcom.js lib to fetch info about domains of the current site and uses that data to run conditions that figure out which notices to show.

### [Stale Cart](https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/current-site/stale-cart-items-notice.js)
- Uses the Flux store (https://github.com/Automattic/wp-calypso/blob/master/client/lib/cart/store/index.js) and shows the notice when there are items in the cart.
- Additional display conditions apply, like showing the notice every 10 minutes when dismissed etc…
- The cart state is synchronized with the API and it also uses the localStorage polling method to sync the state across all opened browser tabs with Calypso
- Doesn't render any notice directly but uses Redux `state/notices` to add a notice to a list of those to display

### [Other Notices](https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/current-site/notice.jsx)
- This it a kitchen sink component of many other types of notices, queries a ton of state — user data (currency code, capabilities), current site, its domains, store discounts, checks for site being "wp for teams" or Jetpack, site plan ownership…

#### Site Redirect Notice
- says that the site is fully redirected to a different URL and displays the target URL
- `Notice` component

#### Active Discount Notice
- checks eligibility of site/user for one of https://github.com/Automattic/wp-calypso/blob/master/client/lib/discounts/active-discounts.js
- Shows `UpsellNudge` with data from the discount definition

# Conclusion
This was just a first pass over the existing nudges and it turns out this is a pretty complicated chunk of code which uses different store types (Redux, Flux) and several different API types for feeding the data

It relies on entities like user, site, domain, store, cart, plan and has no central logic that does all the checks. The eligibility logic is scattered across many different components and their methods. Should we want to achieve 1:1 copy in wp-admin, it will require a huge rewrite of the code. Copying the full logic in wp-admin will be an extreme project too.

### PR DESCRIPTION
**Pull request by cpapazoglou**
Current PR is an attempt to showcase whether we can iframe the `site-switcher`, `site-card`, `nudges` and `notices` within `wp-admin`. It is a very draft - rough prototype so that we can pinpoint feasibility and possible issues.

#### Changes proposed in this Pull Request

* create a page containing only the sidebar and masterbar
* visually hide masterbar, menu items, notices, inline-help

#### Current Issues
- Security - In order for calypso screens to be iframed in inside example.com we need to send a `X-Frame-Options: "example.com"` . Since we may have unlimited domains, I am not sure if / whether we can bypass this. Possible fix 12df840
- Styling - The iframe needs to have variable height (because we may - may not have notices). This can possibly be achieved through JS.
- UX - URLs have to redirect the parent page (`wp-admin`) to calypso (not navigate inside the iframe)

#### Testing instructions

* Go to http://calypso.localhost:3000/site-card/YOURDOMAIN (pick a site that has notices, such as "Free domain with a plan"
* You should something like the following

* Fire up YOURDOMAIN/wp-admin and insert `<iframe src="http://calypso.localhost:3000/site-card/YOURDOMAIN" frameborder="0" width="200px" height="183px"></iframe>` in the admin menu bar you should get something like the following (just set the admin bar width to 200px!)

Resolves #45574

### CODE REVIEW

File: client/my-sites/site-card/index.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+import { navigation, siteSelection } from 'my-sites/controller';
+import { makeLayout, render as clientRender } from 'controller';
+
+/**
+ * Style dependencies
+ */
+import './style.scss';
+
+export default function () {
+	page( '/site-card/:domain?', siteSelection, navigation, null, makeLayout, clientRender );
```
Comment (tyxla) [inline client/my-sites/site-card/index.js:19]: It's worth noting that using this without a site slug or with a wrong site slug will bring up the "All Sites" or site selector (depending on what we render). We should likely have a different approach here, where I'd guess we render nothing in that case.

Comment (tyxla) [inline client/my-sites/site-card/index.js:19]: Did we mean to put something in place of that `null`?

File: client/server/pages/index.js [hunk 2]
```diff
@@ -252,8 +252,9 @@ function setUpLocalLanguageRevisions( req ) {
 }
 
 function setUpLoggedOutRoute( req, res, next ) {
+	// We should change that back to SAMEORIGIN before merging to master
 	res.set( {
-		'X-Frame-Options': 'SAMEORIGIN',
+		'X-Frame-Options': 'allow',
```
[the remark this answers is not in the document]
Comment (cpapazoglou) [reply client/server/pages/index.js:257]: Sure, thanks for pointing out (I have left it as a comment too). I thinks this part is the biggest blocker. If can't figure out how we can bypass it (without setting allow) we are on a dead end on this proposal. It seems to me that we could build a mechanism that sets `X-Frame-Options: YOURDOMAIN` based on the `http://calypso.localhost:3000/site-card/YOURDOMAIN` ?

File: client/sections.js [hunk 3]
```diff
@@ -461,6 +461,12 @@ const sections = [
 			},
 		],
 	},
+	{
+		name: 'site-card',
```
[the remark this answers is not in the document]
Comment (cpapazoglou) [reply client/sections.js:465]: Great suggestion! Will do!

File: client/my-sites/site-card/index.js (hunk 1 above)
Comment (cpapazoglou) [reply client/my-sites/site-card/index.js:19]: The iframe url will be served by the server, so the possibility of a slug being wrong is pretty low but it is a very good catch! Will certainly do so!

Comment (cpapazoglou) [reply client/my-sites/site-card/index.js:19]: That argument loads the main section (primary). `null` is wrong cause it is not a function. I tried 
```
( context, next ) => next()
```
but couldn't get the nudges / notices. Since it is a just a proposal I just skipped it. Will need to fix it though eventually! Forgot to put it in the Current Issues list. Thanks!

File: client/server/pages/index.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (cpapazoglou) [reply client/server/pages/index.js]: I pushed a commit that hopefully can resolve this c0023bd

[the remark this answers is not in the document]
Comment (tyxla) [reply client/server/pages/index.js]: 👍  Thanks for working to address this issue. I'm not sure it resolves the underlying problem, though.

It's important to note that this won't do the job, because `X-Frame-Options` supports only `SAMEORIGIN` and `DENY` currently. The `ALLOW-FROM` directive that allows us to specify a particular host is deprecated and is no longer supported in modern browsers.

Theoretically, we could also use the Content-Security-Policy `frame-ancestors` directive, which allows us to specify a valid parent that would be allowed to iframe a certain route in Calypso. This would actually _work_.

However, this still doesn't solve the main problem, being that is that nothing would stop a potential offensive site from doing the following:
* Connecting a self-hosted WordPress site to WordPress.com - let's call it `https://mybadbadmaliciouswebsite.com`
* Iframe `https://wordpress.com/site-card/mybadbadmaliciouswebsite.com` into their site.
* Setup a simple clickjacking technique on top of the iframe.
* Trick the user into trying to click inside the iframe, while they actually click on the hidden malicious link.

It is even more concerning that with this, the scripts will be able to access the iframe's document and its internal DOM, and use that to manipulate the iframe contents by using `contentWindow`.

So, to summarize, I think that this concept, regardless of how it's set up, always allows potential clickjacking exploits with a certain part of Calypso. Therefore I'd really be hesitant to go with that approach.

Have you considered any alternatives? Using Calypso as a framework to build an app and render that app instead, or using Calypso to build a package that could be consumed instead? These approaches could help reusing Calypso components and logic, while staying on the safe side.

[the remark this answers is not in the document]
Comment (cpapazoglou) [reply client/server/pages/index.js]: >Have you considered any alternatives? Using Calypso as a framework to build an app and render that app instead, or using Calypso to build a package that could be consumed instead? These approaches could help reusing Calypso components and logic, while staying on the safe side.

This PR aimed to avoid going the **long road** of decoupling that component from the calypso state which as you mentioned is a good and valid approach. Since your security concerns actually make sense, I will close this PR and regard this approach as a no go. Thanks a lot for chiming in and your prompt feedback!

**Comment by cpapazoglou** [comment]
We shall not proceed with an iFrame approach due to security concerns 
https://github.com/Automattic/wp-calypso/pull/45780#discussion_r492552101
